keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,435 @@
+* The kubelet will create a cgroup for the pod that expects to get roughly `300m` cores, but the container cgroup created for `copy-files` will request `100m` while that init container is running
+* The quota subsystem will block this pod from being created if there is less than `300m` available.
+
+Once a pod is created, it passes through three high level lifecycle states as seen by the total Kubernetes system. These states are defined in terms of an outside observer of the API - because Kubernetes is a distributed system individual components may report status such as the pod phase after the transition has already occurred, and it is important to clarify what assumptions an API observer may make about those states. The naming is chosen to align with the phases reported by the pod, but we clarify the difference between the Kubelet's view of phase and how that outside observer should view that state without changing the meaning of the pod's status fields.
+
+The first state is `Schedulable` - the time before the pod is scheduled to a node (this is denoted by the nodeName field being set, not by the status update made by the Kubelet to the phase). A pod is `Schedulable` after a successful create via the API. The second state is `Runnable` and occurs immediately after `nodeName` is set, when the Kubelet may begin initializing resources on the node and may start or restart process, and continues until the pod is gracefully deleted via the API, reaches a terminal state of success or failure when `restartPolicy` is `Never` or `OnFailure`, or is rejected by the Kubelet and put in the terminal state of failure. The final state is `Completed` which means the pod has no running containers, all resources have been released or cleaned up, and the pod will never again consume those resources. A pod transitions to the `Completed` terminal state when pod status phase has reached Succeeded or Failed (as observed by the Kubelet and then recorded into the API), or the pod is marked for deletion and all containers are reported stopped via status to the API. A pod in the `Completed` state may be referred to as a "terminal pod" or a "terminated pod". A pod that has been marked for termination via the DELETE API call but not yet reached a terminal state may be referred to as a "gracefully deleted pod" (or "deleted pod" for short) or a "terminating pod". We define one sub state of `Runnable` - the `Initializing` state is the time between when the pod is scheduled on a node and when all init containers have completed successfully and the `Initialized` condition on the pod is set to `true`. A pod without init containers may only remain in this phase very briefly. Because the Kubelet is not required to write pod status before it begins execution of a container, the `Initializing` phase is not consider a distinct top level phase (it is not distiguishable safely from a system perspective) except when presented by the Kubelet directly via metrics or on a node.
```

Comment:
I would say that talking about the "Initializing" sub-state is irrelevant for the purposes of this KEP (the resource model), but you can keep it if you feel very strongly about it.

Runnable is already well defined and that's all we need for resource accounting.

Reply:
Actually, I think i'll just nuke it.  As we've refined the kep and dropped the metric that showed that difference between initializing and running there is no need to clarify here.  We said we might come back and amend the KEP for that in a later release, and defining initializing can be deferred until then.